test/restart.jl

```diff
@@ -0,0 +1,335 @@
+  over all the properties. You can pass a base name.
+- `ignore` is a collection of `Symbol`s that identify properties that are
+  ignored when walking through the tree. This is useful for properties that
+  are known to be different (e.g., `output_dir`).
+
+`:rc` is some CUDA/CuArray internal object that we don't care about
+"""
+function compare(v1, v2; name = "", ignore = Set([:rc]))
+    @testmsg propertynames(v1) == propertynames(v2) name
+
+    properties = filter(x -> !(x in ignore), propertynames(v1))
+
+    if isempty(properties)
+        # Base case
+        @testmsg typeof(v1) == typeof(v2) name
+
+        if v1 isa Field || v1 isa Array || v1 isa AbstractData
+            if eltype(v1) <: AxisTensor
+                # We have to handle the AxisTensor case separately because it
+                # behaves differently
+                for (f, g) in zip(v1, v2)
+                    compare(f, g; name = "$(name)")
+                end
+            elseif eltype(v1) <: CartesianIndex
+                @testmsg v1 == v2 name
+            else
+                arr1, arr2 = Array(parent(v1)), Array(parent(v2))
+                if eltype(arr1) <: AbstractFloat
+                    error = _error(arr1, arr2)
+                    isempty(error) ||
+                        @testmsg maximum(error) <= 100eps(eltype(arr1)) name
+                else
+                    @testmsg arr1 == arr2 name
+                end
+            end
+        elseif Base.issingletontype(typeof(v1))
+            # We have already tested this
+        elseif v1 isa AbstractString || v1 isa Symbol
+            @testmsg v1 == v2 name
+        elseif v1 isa Number
+            # We check with triple equal so that we also catch NaNs being equal
+            @testmsg v1 === v2 name
+        else
+            # Uncaught case, we cannot compare the two
+            @testmsg false name
+        end
+    else
+        # Recursive case
+        for p in properties
+            compare(
+                getproperty(v1, p),
+                getproperty(v2, p);
+                name = "$(name).$(p)",
+            )
+        end
+    end
+end
+
+# Begin tests
+
+# Disable all the @info statements that are produced when creating a simulation
+Logging.disable_logging(Logging.Info)
+
+if comms_ctx isa ClimaComms.SingletonCommsContext
+    configurations = ["sphere", "box", "column"]
+else
+    configurations = ["sphere", "box"]
+end
+
+for configuration in configurations
+    if configuration == "sphere"
+        moistures = ["equil", "nonequil"]
+        precips = ["0M", "1M"]
+        topography = "Earth"
+        turbconv_models = [nothing, "diagnostic_edmfx"]
+        # turbconv_models = ["prognostic_edmfx"]
+    else
+        moistures = ["equil"]
+        precips = ["1M"]
+        topography = "NoWarp"
+        turbconv_models = ["diagnostic_edmfx"]
+    end
+
+    for turbconv_mode in turbconv_models
+        for moisture in moistures
+            for precip in precips
```

Comment:
One problem with this loop is that lots of objects inside the loop will be type-unstable, but this is arguably a problem with how we configure the driver.

I don't know if we for sure need to make changes to this loop, but I did notice that it takes quite a while to run (I ctrl-C'd after a while on my computer).

It could be an issue with allocations inside `compare`, so perhaps it'll be fine after refactoring that.

Using `let` is a bit of a trade-off: using `let` prevents leaky tests, but it also means that we need to compile the same methods multiple times.